pkg/auth/providers/azure/clients/ms_graph_client.go

```diff
@@ -112,11 +114,42 @@ func (c azureMSGraphClient) ListGroupMemberships(id string) ([]string, error) {
 func (c azureMSGraphClient) LoginUser(config *v32.AzureADConfig, credential *v32.AzureADLogin) (v3.Principal, []v3.Principal, string, error) {
 	logrus.Debugf("[%s] Started token swap with AzureAD", providerLogPrefix)
 
-	// Acquire the OID just to verify the user.
-	oid, err := oidFromAuthCode(credential.Code, config)
-	if err != nil {
-		return v3.Principal{}, nil, "", err
+	var oid string
+	if credential.IDToken != "" {
+		issuer := fmt.Sprintf("%s%s/v2.0", config.Endpoint, config.TenantID)
+		provider, err := oidc.NewProvider(context.Background(), issuer)
+		if err != nil {
+			return v3.Principal{}, nil, "", err
+		}
+
+		verifier := provider.Verifier(&oidc.Config{ClientID: config.ApplicationID})
+		token, err := verifier.Verify(context.Background(), credential.IDToken)
+		if err != nil {
+			return v3.Principal{}, nil, "", err
+		}
+
+		type azureClaims struct {
```

Comment:
You can simplify this to:

```go
var claims struct {
        OID string `json:"oid"`
}
```

I'd also be inclined to extract this to `oidFromIDToken(credential.IDToken, config)` similar to `oidFromAuthCode` i.e. "Single Level of Abstraction Principal".

Reply:
Just to clarify my understanding of the code, basically, the CLI gets a "Device Authorization Grant" ID token and sends that as the credentials, if this is provided, we verify, parse, and extract the `oid`, as opposed to getting the OID by hitting the Azure API and getting it.

Reply:
Exactly. In this case the CLI is doing the exchange and we already have the token.

In the other one the server will need to fetch a token exchanging the `code` received by the frontend.

Reply:
Done with adcccb8